lib/puppet/pops/loader/type_definition_instantiator.rb

```diff
@@ -42,14 +42,61 @@ def self.create(loader, typed_name, source_ref, pp_code_string)
     # the loader is known - hence this mechanism
     private_loader = loader.private_loader
     Adapters::LoaderAdapter.adapt(type_definition).loader = private_loader
-
-    Types::PTypeAliasType.new(name, type_definition.type_expr)
+    create_type(type_definition, loader)
   end
 
   def self.create_from_model(type_definition, loader)
+    typed_name = Loader::TypedName.new(:type, type_definition.name.downcase)
+    type = create_type(type_definition, loader)
+    loader.set_entry(
+      typed_name,
+      type,
+      Adapters::SourcePosAdapter.adapt(type_definition).to_uri)
+    type
+  end
+
+  def self.create_type(type_definition, loader)
+    type_expr = type_definition.type_expr
     name = type_definition.name
-    [Loader::TypedName.new(:type, name.downcase), Types::PTypeAliasType.new(name, type_definition.type_expr)]
+    if object_definition?(type_expr)
+      i12n_hash_expr = type_expr.keys[0]
+      defined_name = object_name(i12n_hash_expr)
+      if defined_name.nil? || defined_name == name
+        # No need for an alias. The Object type itself will be receive the name instead
+        Types::PObjectType.new(name, i12n_hash_expr)
+      else
+        # Object type has a name that is different from the alias, it must be bound under that name as well
+        type = Types::PObjectType.new(defined_name, i12n_hash_expr)
```

Comment:
Does this (dual names) imply that $a = Object[{name => foo}] will result in type 'foo' being bound to that Object? (I don't think it should).